sno/pk_generation.py

```diff
@@ -109,16 +128,37 @@ def load_data_from_repo(self, repo):
             for pk, feature_hash in data["generatedPrimaryKeys"].items()
         }
 
-        # Hash of feature contents -> primary key(s) of imported feature(s), for every feature ever imported.
-        self.hash_to_pks = self._invert_pk_map(self.pk_to_hash)
-
-        # Subset of hash_to_pks - only contains primary keys that have not yet been assigned during the current import.
-        # Meaning, if we need a primary key for a feature, we should first check this dict to find a historical one that
-        # hasn't yet been assigned to a feature during this import.
-        self.hash_to_unassigned_pks = self._invert_pk_map(self.pk_to_hash)
-
-        # Next primary key to use if we can't find a historical but unassigned one in hash_to_unassigned_pks.
-        self.next_new_pk = max(self.pk_to_hash) + 1 if self.pk_to_hash else 1
+        # First primary key to use if we can't find a historical but unassigned one.
+        self.first_new_pk = max(self.pk_to_hash) + 1 if self.pk_to_hash else 1
+
+        # The number of inserts, deletes, previous- and current-feature-count, are related by the given formula:
+        # prev-FC + inserts - deletes = curr-FC
+        # Since we know prev-FC and curr-FC already, we can already calculate the number of inserts we can encounter
+        # before we know that (inserts + deletes) definitely exceeds the similarity_detection_limit, and once that many
+        # inserts are encountered, we give up on similarity detection.
+        feature_count_delta = self.feature_count - self.prev_dest_dataset.feature_count
+        self.similarity_detection_insert_limit = (
+            max(self.similarity_detection_limit - feature_count_delta, 0) // 2
+        )
+
+    def _prev_import_dest_tree(self, repo):
+        """Returns the dataset tree that was created the last time this datasource was imported."""
+        if repo.is_empty:
+            return None
+
+        try:
+            # The required dataset tree is in the commit where this file was last modified:
```

Comment:
No way to do this in libgit2/pygit2?

Reply:
There is... it's now done using pygit2. It's about the same amount of code, but I guess it's better not to use git directly if avoidable.